modules/renter/contractor/contractor.go

```diff
@@ -41,27 +41,29 @@ type Contractor struct {
 	hdb     hostDB
 	log     *persist.Logger
 	persist persister
+	mu      sync.RWMutex
 	tpool   transactionPool
 	wallet  wallet
 
-	allowance       modules.Allowance
-	blockHeight     types.BlockHeight
+	// in addition to mu, a separate lock enforces that multiple goroutines
+	// won't try to simultaneously edit the contract set.
+	editLock siasync.TryMutex
+
+	allowance     modules.Allowance
+	blockHeight   types.BlockHeight
+	currentPeriod types.BlockHeight
+	lastChange    modules.ConsensusChangeID
+
+	downloaders map[types.FileContractID]*hostDownloader
+	editors     map[types.FileContractID]*hostEditor
+	renewing    map[types.FileContractID]bool // prevent revising during renewal
+	revising    map[types.FileContractID]bool // prevent overlapping revisions
+
 	cachedRevisions map[types.FileContractID]cachedRevision
 	contracts       map[types.FileContractID]modules.RenterContract
-	currentPeriod   types.BlockHeight
-	downloaders     map[types.FileContractID]*hostDownloader
-	editors         map[types.FileContractID]*hostEditor
-	lastChange      modules.ConsensusChangeID
 	oldContracts    map[types.FileContractID]modules.RenterContract
+	relationships   map[types.FileContractID]types.SiaPublicKey
```

Comment:
well, you can't look anything up in the hostdb without a pubkey because the host only keys things in terms of pubkeys now.

I'm cautious to add anything to `modules.RenterContract` because I'm not sure if its used in the persist anywhere.